### PR TITLE
Update glossary

### PR DESCRIPTION
**Pull request by jkotas**

### CODE REVIEW

**Comment by jkotas** [comment]
@richlander @samsp-msft I have updated the dotnet/runtime to reflect the current state of affairs. I have also included a section on different AOT technologies since it is a common source of confusion. Could you please take a look?

File: docs/project/glossary.md [hunk 1]
```diff
@@ -232,34 +127,58 @@ officially updated since .NET Framework 2.0.
...
+is a file format used by CoreCLR runtime to store AOT compiled code. `crossgen` is the AOT compiler that
+produces binaries in the ReadyToRun file format.
+
+### NGen
+
+[NGen](https://docs.microsoft.com/en-us/dotnet/framework/tools/ngen-exe-native-image-generator)
+is AOT technology included in .NET Framework. It usually compiles code on the machine where
+the code will be executed at install time.
+
+### Full AOT
```
Comment (jkotas) [inline docs/project/glossary.md:166]: @marek-safar Anything to add about the main Mono AOT tech flavors?

File: docs/project/glossary.md [hunk 2]
```diff
@@ -14,9 +14,11 @@ terminology.
...
 | BBT | Microsoft internal early version of C/C++ PGO. See https://www.microsoft.com/windows/cse/bit_projects.mspx. |
 | BOTR | Book Of The Runtime. |
+| BCL | Base Class Library. A set of `System.*` (and to a limited extent `Microsoft.*`) libraries that make up the lower layer of the .NET library stack. |
 | CLR | Common Language Runtime. |
 | COMPlus | An early name for the .NET platform, back when it was envisioned as a successor to the COM platform (hence, "COM+"). Used in various places in the CLR infrastructure, most prominently as a common prefix for the names of internal configuration settings. Note that this is different from the product that eventually ended up being named [COM+](https://msdn.microsoft.com/en-us/library/windows/desktop/ms685978.aspx). |
 | COR | [Common Object Runtime](http://www.danielmoth.com/Blog/mscorlibdll.aspx). The name of .NET before it was named .NET. |
+| CoreFX | Core Framework. Original roject name for open source and cross-platform version of [.NET runtime libraries](https://github.com/dotnet/runtime/tree/master/src/libraries) |
```
Comment (hoyosjs) [inline docs/project/glossary.md:21]:
```suggestion
| CoreFX | Core Framework. Original project name for open source and cross-platform version of [.NET runtime libraries](https://github.com/dotnet/runtime/tree/master/src/libraries) |
```

File: docs/project/glossary.md [hunk 3]
```diff
@@ -232,34 +127,58 @@ officially updated since .NET Framework 2.0.
...
-chunks of the .NET Framework under the MIT license as well, which can be found
-[here][referencesource]. This enabled the Mono community to use the same code
-the .NET Framework uses in order to close gaps and avoid behavioral differences.
+Today, the [Mono runtime](https://github.com/dotnet/runtime/tree/master/src/mono) is part
+of unified .NET platform. It is optimized for mobile (e.g. Xamarin) and browser (e.g. Blazor) scenarios.
+
+### CoreCLR
+
+Originally, CoreCLR was the runtime of Silverlight and was designed to run on multiple
+platforms, specifically Windows and OS X.
```
Comment (hoyosjs) [inline docs/project/glossary.md:141]:
```suggestion
platforms, specifically Windows and macOS.
```

File: docs/project/glossary.md [hunk 4]
```diff
@@ -232,34 +127,58 @@ officially updated since .NET Framework 2.0.
...
-release Rotor as open source, Mono was forced to start from scratch and is thus
+release Rotor as open source, Mono was forced to start from scratch and was thus
 a complete re-implementation of the .NET Framework with no shared code.
 
-When .NET Core was released under the MIT license, Microsoft also released large
-chunks of the .NET Framework under the MIT license as well, which can be found
-[here][referencesource]. This enabled the Mono community to use the same code
-the .NET Framework uses in order to close gaps and avoid behavioral differences.
+Today, the [Mono runtime](https://github.com/dotnet/runtime/tree/master/src/mono) is part
+of unified .NET platform. It is optimized for mobile (e.g. Xamarin) and browser (e.g. Blazor) scenarios.
```
Comment (hoyosjs) [inline docs/project/glossary.md:136]:
```suggestion
of the unified .NET platform. It is optimized for mobile (e.g. Xamarin) and browser (e.g. Blazor) scenarios.
```

File: docs/project/glossary.md (hunk 3 above)
Comment (jkotas) [reply docs/project/glossary.md:141]: The platform name during Silverlight days was OS X. I am tempted to leave it as is.

Comment (hoyosjs) [reply docs/project/glossary.md:141]: I guess it's talking history. That is a fair point then.

File: docs/project/glossary.md [hunk 5]
```diff
@@ -103,114 +82,30 @@ In this document, the following terms are used:
...
-
-### Common Language Runtime
-
-**Also referred to as**: CLR, VM, runtime
-
-The CLR is a virtual machine, i.e. it includes the facilities to generate and
-compile code on-the-fly using a JIT compiler. The existing Microsoft CLR
-implementation is Windows only.
+  executes in. Is also commonly used to refer to the Microsoft .NET Framework
+  Windows only implementation.
```
Comment (stephentoub) [inline docs/project/glossary.md:86]:
```suggestion
  executes in. It is also commonly used to refer to the Microsoft .NET Framework
  Windows-only implementation.
```

File: docs/project/glossary.md [hunk 6]
```diff
@@ -103,114 +82,30 @@ In this document, the following terms are used:
...
-In contrast to the CLR/CoreCLR, CoreRT is not a virtual machine, i.e. it doesn't
-include the facilities to generate and run code on-the-fly because it doesn't
-include a JIT. It does, however, include the GC and the ability for runtime type
-identification (RTTI) as well as reflection. However, its type system is
-designed so that metadata for reflection can be omitted. This enables having an
-AOT tool chain that can link away superfluous metadata and (more importantly)
-identify code that the application doesn't use.
-
-## Platforms
+.NET Core has become future of the platform and we refer to it just as .NET today.
```
Comment (stephentoub) [inline docs/project/glossary.md:96]:
```suggestion
.NET Core has become future of the platform, and we refer to it simply as .NET today. 
```

File: docs/project/glossary.md [hunk 7]
```diff
@@ -103,114 +82,30 @@ In this document, the following terms are used:
...
 ### .NET Framework
 
-**Also referred to as**: Desktop, full framework, in-box framework, ~~CLR~~
+**Also referred to as**: Desktop, full framework
 
-This refers to the .NET Framework that first shipped in 2002 and has been
-updated on a regular basis since then. It's the main framework folks target
-today and allows you to build a wide variety of applications, such as WinForms,
-WPF, ASP.NET, and command line tools.
+.NET Framework was the very first .NET runtime. It first shipped in 2002 and it has been
```
Comment (stephentoub) [inline docs/project/glossary.md:103]:
```suggestion
.NET Framework was the very first .NET runtime. It first shipped in 2002, and it has been
```

File: docs/project/glossary.md [hunk 8]
```diff
@@ -232,34 +127,58 @@ officially updated since .NET Framework 2.0.
...
+Originally, CoreCLR was the runtime of Silverlight and was designed to run on multiple
+platforms, specifically Windows and OS X.
+
+Today, the [CoreCLR runtime](https://github.com/dotnet/runtime/tree/master/src/coreclr) 
+is part of unified .NET platform. It is optimized for cloud (e.g. ASP.NET) and
+desktop (e.g. WinForms, WPF) scenarios.
+
+## Ahead-Of-Time Compilation (AOT)
+
+Most flavors of .NET runtime come with at least partial AOT compilation. Variety of AOT technologies
```
Comment (stephentoub) [inline docs/project/glossary.md:149]:
```suggestion
Most flavors of .NET runtime come with at least partial AOT compilation. A variety of AOT technologies 
```

File: docs/project/glossary.md [hunk 9]
```diff
@@ -232,34 +127,58 @@ officially updated since .NET Framework 2.0.
...
+
+Most flavors of .NET runtime come with at least partial AOT compilation. Variety of AOT technologies 
+with unique characteristics were developed for .NET runtimes over the years.
+
+### ReadyToRun
+
+**Also referred to as**: R2R
+
+[ReadyToRun](https://github.com/dotnet/runtime/blob/master/docs/design/coreclr/botr/readytorun-overview.md)
+is a file format used by CoreCLR runtime to store AOT compiled code. `crossgen` is the AOT compiler that
```
Comment (stephentoub) [inline docs/project/glossary.md:157]:
```suggestion
is a file format used by the CoreCLR runtime to store AOT compiled code. `crossgen` is the AOT compiler that
```

File: docs/project/glossary.md [hunk 10]
```diff
@@ -232,34 +127,58 @@ officially updated since .NET Framework 2.0.
...
+produces binaries in the ReadyToRun file format.
+
+### NGen
+
+[NGen](https://docs.microsoft.com/en-us/dotnet/framework/tools/ngen-exe-native-image-generator)
+is AOT technology included in .NET Framework. It usually compiles code on the machine where
+the code will be executed at install time.
```
Comment (stephentoub) [inline docs/project/glossary.md:164]:
```suggestion
is AOT technology included in .NET Framework. It usually compiles code at install time on the machine where
the code will be executed.
```

File: docs/project/glossary.md [hunk 11]
```diff
@@ -232,34 +127,58 @@ officially updated since .NET Framework 2.0.
...
-necessary to reduce the footprint, both by shipping smaller versions of the
-system libraries as well as making them more linker friendly. While Mono runs on
-macOS without Xamarin, their linker is required make the app package for the
-Mac App Store.  Xamarin ships a full static compiler on iOS, as the platform
-does not support dynamic code generation.
+### Native AOT
 
+[Native AOT](https://github.com/dotnet/designs/blob/main/accepted/2020/form-factors.md#native-aot-form-factors) is
+a .NET runtime form factor with key performance characteristics (startup time, binary size and steady state throughput) competitive
+with statically compiled languages. .NET runtime flavor based on CoreCLR with these characteristics is being developed as
```
Comment (stephentoub) [inline docs/project/glossary.md:180]:
```suggestion
with statically compiled languages. A .NET runtime flavor based on CoreCLR with these characteristics is being developed as
```

**Review by stephentoub — APPROVED** [review]

File: docs/project/glossary.md [hunk 12]
```diff
@@ -14,9 +14,11 @@ terminology.
...
+| BCL | Base Class Library. A set of `System.*` (and to a limited extent `Microsoft.*`) libraries that make up the lower layer of the .NET library stack. |
 | CLR | Common Language Runtime. |
 | COMPlus | An early name for the .NET platform, back when it was envisioned as a successor to the COM platform (hence, "COM+"). Used in various places in the CLR infrastructure, most prominently as a common prefix for the names of internal configuration settings. Note that this is different from the product that eventually ended up being named [COM+](https://msdn.microsoft.com/en-us/library/windows/desktop/ms685978.aspx). |
 | COR | [Common Object Runtime](http://www.danielmoth.com/Blog/mscorlibdll.aspx). The name of .NET before it was named .NET. |
+| CoreFX | Core Framework. Original project name for open source and cross-platform version of [.NET runtime libraries](https://github.com/dotnet/runtime/tree/master/src/libraries) |
 | DAC | Data Access Component. An abstraction layer over the internal structures in the runtime. |
 | EE | Execution Engine. |
 | GC | [Garbage Collector](https://github.com/dotnet/runtime/blob/master/docs/design/coreclr/botr/garbage-collection.md). |
```
Comment (marek-safar) [inline docs/project/glossary.md:24]: Add One .NET ?

File: docs/project/glossary.md [hunk 13]
```diff
@@ -232,34 +127,58 @@ officially updated since .NET Framework 2.0.
...
+[Full AOT](https://docs.microsoft.com/en-us/xamarin/ios/internals/architecture) is used
+by Mono runtime in environments that prohibit fallback to JIT.
 
-Mono has implementations (though not necessarily complete) of WinForms, ASP.NET,
-and System.Drawing.
+### Hybrid AOT
 
-### Xamarin
+[Hybrid AOT](https://docs.microsoft.com/en-us/xamarin/mac/internals/aot#hybrid-aot) is used
+by Mono runtime in environment that allow fallback to JIT.
```
Comment (jkotas) [inline docs/project/glossary.md:174]:
```suggestion
by Mono runtime in environments that allow fallback to JIT. 
```

File: docs/project/glossary.md [hunk 14]
```diff
@@ -232,34 +127,58 @@ officially updated since .NET Framework 2.0.
...
 the same time the .NET Framework was first released. Since Microsoft didn't
-release Rotor as open source, Mono was forced to start from scratch and is thus
+release Rotor as open source, Mono was forced to start from scratch and was thus
 a complete re-implementation of the .NET Framework with no shared code.
 
-When .NET Core was released under the MIT license, Microsoft also released large
-chunks of the .NET Framework under the MIT license as well, which can be found
-[here][referencesource]. This enabled the Mono community to use the same code
-the .NET Framework uses in order to close gaps and avoid behavioral differences.
+Today, the [Mono runtime](https://github.com/dotnet/runtime/tree/master/src/mono) is part
```
Comment (marek-safar) [inline docs/project/glossary.md:135]:
```suggestion
Today, the [Mono VM](https://github.com/dotnet/runtime/tree/master/src/mono) is part
of the unified One .NET platform. It is optimized embeddable VM used primary for mobile, browser or gaming like scenarios.

File: docs/project/glossary.md [hunk 15]
```diff
@@ -103,114 +82,30 @@ In this document, the following terms are used:
...
+### .NET Core / .NET 
 
-It's built from the same code base as the CLR. Originally, CoreCLR was the
-runtime of Silverlight and was designed to run on multiple platforms,
-specifically Windows and OS X. CoreCLR is now part of .NET Core and represents a
-simplified version of the CLR. It's still a [cross platform][core-build-status]
-runtime. CoreCLR is also a virtual machine with a JIT.
+.NET Core has been the name for the open source, cross-platform stack that 
+ASP.NET Core and UWP applications are built on. For more details, 
+read [this blog post][introducing-net-core].
```
Comment (gfoidl) [inline docs/project/glossary.md:94]: Can we name the link title? `this blog post` is quite generic...
```suggestion
read [Introducing .NET Core][introducing-net-core].
```

File: docs/project/glossary.md [hunk 16]
```diff
@@ -103,114 +82,30 @@ In this document, the following terms are used:
...
-include the facilities to generate and run code on-the-fly because it doesn't
-include a JIT. It does, however, include the GC and the ability for runtime type
-identification (RTTI) as well as reflection. However, its type system is
-designed so that metadata for reflection can be omitted. This enables having an
-AOT tool chain that can link away superfluous metadata and (more importantly)
-identify code that the application doesn't use.
-
-## Platforms
+.NET Core has become future of the platform, and we refer to it just as .NET today. 
+For more details, read [this blog post][introducing-net-5].
```
Comment (gfoidl) [inline docs/project/glossary.md:97]:
```suggestion
For more details, read [Introducing .NET 5][introducing-net-5].
```

File: docs/project/glossary.md [hunk 17]
```diff
@@ -232,34 +127,58 @@ officially updated since .NET Framework 2.0.
...
+## Ahead-Of-Time Compilation (AOT)
+
+Most flavors of .NET runtime come with at least partial AOT compilation. Variety of AOT technologies 
+with unique characteristics were developed for .NET runtimes over the years.
+
+### ReadyToRun
+
+**Also referred to as**: R2R
+
+[ReadyToRun](https://github.com/dotnet/runtime/blob/master/docs/design/coreclr/botr/readytorun-overview.md)
```
Comment (marek-safar) [inline docs/project/glossary.md:156]:
```suggestin
[ReadyToRun](../design/coreclr/botr/readytorun-overview.md)
```

File: docs/project/glossary.md (hunk 1 above)
Comment (marek-safar) [reply docs/project/glossary.md:166]: I think Full AOT is same as NativeAOT in this context

File: docs/project/glossary.md [hunk 18]
```diff
@@ -232,34 +127,58 @@ officially updated since .NET Framework 2.0.
...
-to other platforms, see [Mono's Supported Platforms][mono-supported-platforms]
+[Full AOT](https://docs.microsoft.com/en-us/xamarin/ios/internals/architecture) is used
+by Mono runtime in environments that prohibit fallback to JIT.
 
-Mono has implementations (though not necessarily complete) of WinForms, ASP.NET,
-and System.Drawing.
+### Hybrid AOT
 
-### Xamarin
+[Hybrid AOT](https://docs.microsoft.com/en-us/xamarin/mac/internals/aot#hybrid-aot) is used
```
Comment (marek-safar) [inline docs/project/glossary.md:173]: [Hybrid AOT](https://docs.microsoft.com/en-us/xamarin/mac/internals/aot#hybrid-aot) is used
by Mono runtime in the environment that allows fallback to JIT or environment which requires read-execute protection and still need to execute IL code at runtime.

File: docs/project/glossary.md (hunk 12 above)
Comment (jkotas) [reply docs/project/glossary.md:24]: One .NET was our internal acronym for some time. I do not think it was used much in any public communication, if at all. For example, it was never described in any blog posts as name of a thing.

**Comment by jkotas** [comment]
> I think Full AOT is same as NativeAOT in this context

It is not. Full AOT alone (in Mono at least) does not provide the performance characteristics of Native AOT.

File: docs/project/glossary.md (hunk 14 above)
Comment (jkotas) [reply docs/project/glossary.md:135]: > embeddable 

I think I would need to define what it means to be "embeddable" somewhere.

> gaming 

I believe that the only runtime optimized for gaming is Unity. I will mention Unity here. Outside Unity, I do not think we really heave runtime optimized for gaming. Games outside Unity use every other .NET runtime out there, including niche runtimes like BRUTE.

**Comment by marek-safar** [comment]
> It is not. Full AOT alone (in Mono at least) does not provide the performance characteristics of Native AOT.

I don't follow. Which performance characteristics are unique to native AOT? Should that be the only differentiate? I think that could make it very hard for us to explain especially as such characteristics are measured relatively.

File: docs/project/glossary.md (hunk 14 above)
Comment (marek-safar) [reply docs/project/glossary.md:135]: I cannot list specifics but we have been offering this to our paying customers as official SDKs so Unity is only one of the offerings.

**Comment by jkotas** [comment]
We are discussing more potential updates offline.

**Pull request merged**